Update README, CONTRIBUTING, and other docs (#821)
IntersectMBO/formal-ledger-specifications: CONTRIBUTING.md

Reviewer: The directory structure given in the document does not correspond to reality. I would scrap it altogether.

**Brief Contents**

- [Style Guidelines](#style-guidelines)
- [Project Overview](#project-overview)
- [Development Environment Setup](#development-environment-setup)
- [Building Project Artifacts](#building-project-artifacts)
- [IDE Integration](#ide-integration)
- [CI/CD Workflow](#cicd-workflow)
- [Setup Without Nix (Advanced)](#setup-without-nix-advanced)
- [Miscellanea](#miscellanea)
- [Maintainers](#maintainers)

---

## Style Guidelines

We adhere to the [Agda standard library style
guide](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md) where
practical. However, because our code is *literate* Agda which is used to produce html
documentation, readability of the latter takes precedence over code formatting, so
deviations are expected.

---

## Project Overview

This repository uses [Nix](https://nixos.org/) to provide a reproducible, declarative, and portable development environment.

### Nix and Flakes

> We use Nix to manage all dependencies, build processes, and development shells. The modern [Flakes](https://nixos.wiki/wiki/Flakes) feature is the recommended interface for interacting with the project.
> 

Reviewer: ```suggestion
We use Nix to manage all dependencies, build processes, and development shells. [Nix Flakes](https://nixos.wiki/wiki/Flakes) are the recommended interface for interacting with the project.

```

### Dependency Management

> All external Nix dependencies (like `nixpkgs`) and Agda libraries are pinned to specific versions or Git commits using [niv](https://github.com/nmattia/niv). The pinned versions are stored in `build-tools/nix/sources.json`, which guarantees reproducible builds.

Reviewer: ```suggestion
All external Nix dependencies (like `nixpkgs`) and Agda libraries are pinned to specific versions (e.g., Git commits) using [niv](https://github.com/nmattia/niv). The pinned versions are stored in `build-tools/nix/sources.json`, which guarantees reproducible builds.
```

        └── nix/fls-agda.nix                # Custom FLS Agda derivation
```

---

## Development Environment Setup

We provide several development shells tailored for different tasks. You can enter them using `nix develop`.

### Default Shell

This is the primary environment for Agda development. It includes Agda, all required libraries, and the `fls-shake` build tool.

```bash
# Enter the default development shell
nix develop
```

**Available Tools**

+ `agda` (with all project libraries)
+ `fls-shake` (custom build tool)
+ `python311`
+ `hpack`

### CI Shell

A minimal environment designed for automated builds, containing only the `fls-shake` build tool and its runtime dependencies.

```bash
# Enter the CI shell
nix develop .#ci
```

### Documentation Shell

A comprehensive environment for working on documentation, including the full documentation generation pipeline.

```bash
# Enter the documentation shell
nix develop .#docs
```

**Available Tools:**

* Everything from the default shell.
* `pandoc`, a full `latex` distribution.
* `mkdocs` and its Python dependencies.

> * `mdbook`, `cargo`, and related tools.

Reviewer: I did not understand how this list is organized.

Author: Good catch.  (fixed)

-----

## Building Project Artifacts

> You can build project artifacts in several ways. The recommended method is using `nix build`.

Reviewer: I would expect that the recommended method corresponds to one of the subsections below

Author: Yes, good point.  I think this is left over from when it said something like "the recommended way is with `nix build ...`, but you can also use `nix-build -A ...`."  I'll remove that sentence.
